Openwhisk in a standalone runnable jar

chetanmeh opened the pull request:
Enables running a minimal OpenWhisk (with in memory store) as a runnable jar outside of docker

## Description

Currently users looking to explore  OpenWhisk directly have few options like [devtools docker compose][1], mini kube etc. This PR provides yet another option which enables launching a fully working OpenWhisk server via a runnable jar

```
java -jar openwhisk-standalone-1.0.0-SNAPSHOT.jar
```

This launches a fully functional OpenWhisk controller and invoker. You can run `wsk` commands against it and launch and execute actions etc. 

<img width="617" alt="image" src="https://user-images.githubusercontent.com/664531/60118122-3cf52500-9799-11e9-9241-dc00fca94c20.png">

To try it out currently you can download a pre built jar from [here](https://github.com/chetanmeh/incubator-openwhisk/releases/tag/v0.10) 

### Design

This PR builds up on [lean openwhisk](#4216) support. It configures OpenWhisk to use a `MemoryArtifactStore` (thus avoiding nead of a separate CouchDB container) and enables executing the standard `wsk` commands against it

See also the [discussion thread](https://lists.apache.org/thread.html/7425131f1fc11a9fd21e9c049be702837841c47004da03b7f215a0d6@%3Cdev.openwhisk.apache.org%3E)

Key aspects

* Configured by default to use a `MemoryArtifactStore`
* Can be configured to use another store via simple config
* Delievered as a jar executable
* Adapts the config as per OS
* Bootstraps a set of default users (guest and whisk.system)

### Usage

First build the standalone module

```bash
./gradlew :core:standalone:build
```

*In future*
```
wget maven/openwhisk.jar
```

Launch the server
```
$ java -jar openwhisk-standalone-1.0.0-SNAPSHOT.jar -h
OpenWhisk standalone launcher
  -c, --config-file  <arg>   application.conf which overwrites the default
                             whisk.conf
  -m, --manifest  <arg>      Manifest json defining the supported runtimes
  -p, --port  <arg>          Server port
  -h, --help                 Show help message
$ java -jar openwhisk-standalone-1.0.0-SNAPSHOT.jar
```

### When to use

This mode may be benefical in cases like

* Running simple test against an OpenWhisk server say when testing compositions
* Faster development cycle - No need to build and relaunch docker container to test out simple changes in regular development flow on OpenWhisk core codebase
* Easier entrypoint for new users to get a feel of OpenWhisk features

## Related issue and scope
<!--- Please include a link to a related issue if there is one. -->
- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components
<!--- Select below all system components are affected by your change. -->
<!--- Enter an `x` in all applicable boxes. -->
- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes
<!--- What types of changes does your code introduce? Use `x` in all the boxes that apply: -->
- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:
<!--- Please review the points below which help you make sure you've covered all aspects of the change you're making. -->

- [ ] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [ ] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

[1]: https://github.com/apache/incubator-openwhisk-devtools/tree/master/docker-compose

chetanmeh commented:
This is an initial version to seek feedback on approach. Based on feedback we can decide what all aspects to support with this approach

rabbah commented:
cool!

bdelacretaz commented:
Big +1 to this, great work! I do have a use case where we want to simulate OpenWhisk for local development of actions, looks like this would help a lot.

rabbah commented on common/scala/src/main/scala/org/apache/openwhisk/core/database/memory/MemoryArtifactStore.scala:
```diff
@@ -285,8 +296,10 @@ class MemoryArtifactStore[DocumentAbstraction <: DocumentSerializer](dbName: Str
   override def shutdown(): Unit = {
-    artifacts.clear()
-    attachmentStore.shutdown()
+    if (!ignoreShutdown) {
```
worth commenting this - i dont understand why it would be necessary at face value.

rabbah commented:
Can we add a test for this in Jenkins?

chetanmeh commented:
There are still some pending issues like getting log collection to work on Mac. For now implementing an approach which uses `docker logs` command to fetch logs. 

With that change the system test pass percentage improved

> 108 tests completed, 14 failed

chetanmeh commented:
> Can we add a test for this in Jenkins? Like with lean - stand it up, create/run/delete a single action as a sniff test.

Yeah working on that. Its bit tricky to get it working in `tests` project as it needs some different default global config. For now only 14 out of 108 system tests fail

starpit commented:
invoking composer compositions don't seem to work, due to a lack of the `redis` npm

dgrove-oss commented:
@tardieu  -- see @starpit's comment above.  Any advice for @chetanmeh on how to fix?

starpit commented:
oh, i think i fixed this one on my own, by specifying a custom manifest. i guess the default manifest uses a very slim nodejs container (guessing).

starpit commented:
after wasting a fair amount of time on this... this seems a bit buggy to me, still. e.g. if i deploy a Composer composition, invoke it; then delete the action and re-create it with a different compostion... subsequent GETs on the composer action give the expected update, but INVOKEs stick with the old verson -- until i restart the openwhisk java process.

here is what i see from the INVOKE, which seems to indicate that the cached container has not been invalidated:
```
[WhiskAction] [GET] serving from cache: CacheKey(guest/hhh) [marker:database_cacheHit_counter:10]
[ContainerPool] containerStart containerState: warmed container: Some(ContainerId(45ce6c4b5a90bffb69a164ca32358fb5f8372e98a7c8a0c75238a3bbc43c2200)) activations: 1 of max 1 action: hhh namespace: guest activationId: 729a96fe2a2140109a96fe2a21401078 [marker:invoker_containerStart.warmed_counter:10]

```

here is the detail of the DELETE:
```
DELETE /api/v1/namespaces/_/actions/hhh 
[GET] serving from cache: CacheKey(23bc46b1-71f6-4ed5-8c54-816aa4f8c502) [marker:database_cacheHit_counter:1]
[WhiskAction] [GET] serving from cache: CacheKey(guest/hhh) [marker:database_cacheHit_counter:2]
[WhiskAction] invalidating CacheKey(guest/hhh) on delete
 [MemoryArtifactStore] [DEL] 'whisks' deleting document: 'id: guest/hhh, rev: 1' [marker:database_deleteDocument_start:2]
[MemoryArtifactStore]  [marker:database_deleteDocument_finish:2:0]
[WhiskAction] invalidating CacheKey(guest/hhh)
[MemoryAttachmentStore] [ATTS_DELETE] uploading attachment of document 'guest/hhh' [marker:database_deleteDocumentAttachments_start:6]
[MemoryAttachmentStore]  [marker:database_deleteDocumentAttachments_finish:6:0]
[BasicHttpService] [marker:http_delete.200_counter:7:7]
```

and the CREATE:
```
PUT /api/v1/namespaces/_/actions/hhh 
[WhiskAction] [GET] serving from datastore: CacheKey(guest/hhh) [marker:database_cacheMiss_counter:5]
[MemoryArtifactStore] [GET] 'whisks' finding document: 'id: guest/hhh' [marker:database_getDocument_start:5]
[MemoryArtifactStore]  [marker:database_getDocument_finish:5:0]
[WhiskAction] invalidating CacheKey(guest/hhh)
[WhiskAction] write initiated on new cache entry
[MemoryArtifactStore] [PUT] 'whisks' saving document: 'id: guest/hhh, rev: 0' [marker:database_saveDocument_start:7]
[MemoryArtifactStore]  [marker:database_saveDocument_finish:7:0]
[WhiskAction] write all done, caching CacheKey(guest/hhh) Cached
```

rabbah commented:
@starpit does the same happen for an action that's created/invoked/deleted/recreated/invoked?

starpit commented:
indeed. but i think you have to do a delete followed by a create; for some reason updates do not cause this to occur.

and openwhisk-composer happens to do a delete followed by create, always -- probably because @tardieu wanted to make sure annotations were invalidated (because they aren't, on update)

rabbah commented:
I suspect I know the issue - if an action is identical but the only
difference is the document revision they are considered equal and the
invoker will reuse the container.
https://github.com/apache/incubator-openwhisk/issues/3120

starpit commented:
in my cases (both composer and plain actino), i am actually changing the action source

rabbah commented:
ok then that throws out my suspicion. I'll have a peek I've been meaning to try this feature out.

starpit commented:
in this session, foo1.js is x=>x, and foo2 is x=>x+1; note how updates are fine, but delete followed by create is not

![Screen Shot 2019-06-21 at 20 47 50](https://user-images.githubusercontent.com/4741620/59957386-e66cab80-9465-11e9-8ad2-94f10e625c04.png)

chetanmeh commented:
I think it's due to simple revision strategy in MemoryArtifactStore implement which just increments a number. This gets reset post delete. I would change it to be based on content hash. That should fix the issue then

chetanmeh commented:
@starpit Just fix the issue in `MemeoryArtifactStore`. Can you give it a try again

chetanmeh commented:
> oh, i think i fixed this one on my own, by specifying a custom manifest. i guess the default manifest uses a very slim nodejs container (guessing).

@starpit This is also now fixed in latest. It uses same runtime manifest as used by anisble setup

starpit commented:
now create+delete+create+invoke gives the expected result!

i'll throw this at the kui CI/CD again to see what happens.

starpit commented:
i'm seeing the log end markers showing up, as visible to the user, e.g. from an activation get

```
"logs": [
        "Entering composition",
        "Entering composition.test",
        "Entering composition.consequent",
        "Entering composition.consequent[0]",
        "Entering final state",
        "{\"path\":true}",
        "XXX_THE_END_OF_A_WHISK_ACTIVATION_XXX"
    ],
```

starpit commented:
there are ther differences, e.g. stdout versus stderr distniction in the logs; this is what i see from ibm cloud functions for an activation of the same function

```
  "logs": [
    "2019-06-23T00:20:49.529797Z    stdout: Entering composition",
    "2019-06-23T00:20:49.530068Z    stdout: Entering composition.test",
    "2019-06-23T00:20:49.530238Z    stdout: Entering composition.consequent",
    "2019-06-23T00:20:49.530246Z    stdout: Entering composition.consequent[0]",
    "2019-06-23T00:20:49.530514Z    stdout: Entering final state",
    "2019-06-23T00:20:49.530559Z    stdout: {\"path\":true}"
  ],
```

chetanmeh commented:
> i'm seeing the log end markers showing up, as visible to the user

There was a pending todo for this. Fixed this now

chetanmeh commented:
Uploaded a locally build copy of the standalone server [here][1] so that someone can now try this out quickly.

Test results from system basic test suite

```
./gradlew :tests:testSystemBasic -Dwhisk.auth="23bc46b1-71f6-4ed5-8c54-816aa4f8c502:123zO3xZCLrMN6v2BKK1dXYFpXlPkccOFqm12CdAsMgRU4VrNZ9lyGVCGuMDGIwP" -Dwhisk.server=http://localhost:3233
```

1. On Linux - Most test except the unicode related tests passed
2. On Mac around 16 test fail intermittently. I suspect that is due to instability in connecting to docker container from outside of Docker. Same test pass with ansible setup where Invokers uses direct Docker networking to connect to launched containers. But when running the m via standalone some errors are seem stating that container host cannot be reached

Would now add a basic integration test for standalone case

[1]: https://github.com/chetanmeh/incubator-openwhisk/releases/tag/v0.10

neerajmangal commented:
Hi @chetanmeh , 

First of all thanks for this awesome work!! It's going to help out a lot. 

While testing it I find one issue. 

## Steps 

- NodeJS runtime image is not available on Host `docker images |grep action-nodejs-v10`.
- Start standalone Openwhisk, `java -jar openwhisk-standalone.jar`
- If `action-nodejs-v10` image pull failed during initialization of startup, no warm-up container will be started. 
- Start a `watch docker ps -a` in another terminal. You will not see any warmup container running.
- Pull image of nodejs runtime `docker pull openwhisk/openwhisk/action-nodejs-v10:latest`
- Create nodejs sample action action
- Invoke it, it works first time and you can see the warmup docker instance running in another terminal
- Wait for docker unpause to happen (after 10 mins)
- Again invoke action, it fails with below error.

```bash
wsk action invoke hello -p name neeraj -r -b 
{
    "error": "The action did not initialize and exited unexpectedly."
}

``` 

## Logs 

1. No docker image is available on host machine. 

```bash
(base) Neerajs-MacBook-Pro:openwhisk-standalone mangal$ docker images |grep action-nodejs
openwhisk/action-nodejs-v8                                                        ab224ab                        061e9a2f8dd2        6 weeks ago         959MB
openwhisk/action-nodejs-v12                                                       ab224ab                        bb87dfe9cb8f        4e262c5cede3        6 weeks ago         963MB
openwhisk/action-nodejs-v10                                                       ab224ab                        4e262c5cede3        6 weeks ago         963MB
```

2. Startup logs of the standalone jar. You can see the image pull is failed and there will be no warm container for nodejs runtime in another terminal in `docker ps -a`. 

```bash
java -jar openwhisk-standalone.jar

        ____      ___                   _    _ _     _     _
       /\   \    / _ \ _ __   ___ _ __ | |  | | |__ (_)___| | __
  /\  /__\   \  | | | | '_ \ / _ \ '_ \| |  | | '_ \| / __| |/ /
 /  \____ \  /  | |_| | |_) |  __/ | | | |/\| | | | | \__ \   <
 \   \  /  \/    \___/| .__/ \___|_| |_|__/\__|_| |_|_|___/_|\_\
  \___\/ tm           |_|

Git Commit: 0931083, Build Date: 2019-06-24T14:35:12+0530
================================================================================
Running pre flight checks ...

[  OK   ] 'docker' cli found.
[  OK   ] 'docker' is running.
[  OK   ] 'wsk' cli found.
[  OK   ] 'wsk' configured for namespace [guest].
[  OK   ] 'wsk' configured to connect to http://localhost:3233.
================================================================================
[2019-06-24T17:22:24.154Z] [INFO] Starting OpenWhisk standalone on port 3233
[2019-06-24T17:22:25.520Z] [INFO] Slf4jLogger started
[2019-06-24T17:22:26.836Z] [INFO] [#tid_sid_unknown] [MemoryArtifactStore] Created MemoryStore for [subjects]
[2019-06-24T17:22:26.978Z] [INFO] [#tid_sid_cli] [MemoryArtifactStore] [GET] 'subjects' finding document: 'id: guest' [marker:database_getDocument_start:479]
[2019-06-24T17:22:26.986Z] [INFO] Starting the Kamon(SystemMetrics) module
[2019-06-24T17:22:27.890Z] [INFO] [#tid_sid_cli] [MemoryArtifactStore]  [marker:database_getDocument_finish:1393:1]
[2019-06-24T17:22:27.949Z] [INFO] [#tid_sid_cli] [MemoryArtifactStore] [PUT] 'subjects' saving document: 'id: guest, rev: null' [marker:database_saveDocument_start:1453]
[2019-06-24T17:22:27.956Z] [INFO] [#tid_sid_cli] [MemoryArtifactStore]  [marker:database_saveDocument_finish:1460:7]
[2019-06-24T17:22:27.957Z] [INFO] [#tid_userBootstrap] [StandaloneOpenWhisk] Created user [guest]
[2019-06-24T17:22:28.293Z] [INFO] [#tid_sid_cli] [MemoryArtifactStore] [GET] 'subjects' finding document: 'id: whisk.system' [marker:database_getDocument_start:251]
[2019-06-24T17:22:28.294Z] [INFO] [#tid_sid_cli] [MemoryArtifactStore]  [marker:database_getDocument_finish:251:0]
[2019-06-24T17:22:28.299Z] [INFO] [#tid_sid_cli] [MemoryArtifactStore] [PUT] 'subjects' saving document: 'id: whisk.system, rev: null' [marker:database_saveDocument_start:257]
[2019-06-24T17:22:28.300Z] [INFO] [#tid_sid_cli] [MemoryArtifactStore]  [marker:database_saveDocument_finish:258:1]
[2019-06-24T17:22:28.301Z] [INFO] [#tid_userBootstrap] [StandaloneOpenWhisk] Created user [whisk.system]
[2019-06-24T17:22:28.658Z] [INFO] The kamon.reporters setting is empty, no reporters have been started.
[2019-06-24T17:22:29.182Z] [INFO] [#tid_sid_unknown] [Config] system set value for limits.triggers.fires.perMinute
[2019-06-24T17:22:29.183Z] [INFO] [#tid_sid_unknown] [Config] system set value for limits.actions.sequence.maxLength
[2019-06-24T17:22:29.183Z] [INFO] [#tid_sid_unknown] [Config] system set value for limits.actions.invokes.concurrent
[2019-06-24T17:22:29.183Z] [INFO] [#tid_sid_unknown] [Config] system set value for controller.instances
[2019-06-24T17:22:29.184Z] [INFO] [#tid_sid_unknown] [Config] system set value for whisk.api.host.name
[2019-06-24T17:22:29.184Z] [INFO] [#tid_sid_unknown] [Config] system set value for limits.actions.invokes.perMinute
[2019-06-24T17:22:29.185Z] [INFO] [#tid_sid_unknown] [Config] system set value for whisk.api.host.proto
[2019-06-24T17:22:29.185Z] [INFO] [#tid_sid_unknown] [Config] system set value for whisk.api.host.port
[2019-06-24T17:22:29.185Z] [INFO] [#tid_sid_unknown] [Config] system set value for runtimes.manifest
[2019-06-24T17:22:29.186Z] [INFO] [#tid_sid_unknown] [Config] system set value for port
[2019-06-24T17:22:29.823Z] [INFO] [#tid_sid_unknown] [LeanMessagingProvider] topic completedstandalone created
[2019-06-24T17:22:29.823Z] [INFO] [#tid_sid_unknown] [LeanMessagingProvider] topic health created
[2019-06-24T17:22:29.823Z] [INFO] [#tid_sid_unknown] [LeanMessagingProvider] topic cacheInvalidation created
[2019-06-24T17:22:29.824Z] [INFO] [#tid_sid_unknown] [LeanMessagingProvider] topic events created
[2019-06-24T17:22:30.289Z] [INFO] [#tid_sid_controller] [Controller] starting controller instance standalone [marker:controller_startup_counter:3951]
[2019-06-24T17:22:30.403Z] [INFO] [#tid_sid_unknown] [MemoryArtifactStore] Created MemoryStore for [whisks]
[2019-06-24T17:22:30.455Z] [INFO] [#tid_sid_dispatcher] [MessageFeed] handler capacity = 128, pipeline fill at = 128, pipeline depth = 256
[2019-06-24T17:22:30.513Z] [INFO] [#tid_sid_dispatcher] [MessageFeed] handler capacity = 128, pipeline fill at = 128, pipeline depth = 256
[2019-06-24T17:22:30.915Z] [INFO] [#tid_sid_unknown] [ExtendedDockerClient] Detected docker client version 18.05.0-ce
[2019-06-24T17:22:30.916Z] [INFO] [#tid_sid_unknown] [InvokerReactive] LogStoreProvider: class org.apache.openwhisk.core.containerpool.docker.DockerCliLogStore
[2019-06-24T17:22:31.145Z] [INFO] [#tid_sid_unknown] [DockerForMacClient] Detected docker client version 18.05.0-ce
[2019-06-24T17:22:31.178Z] [INFO] [#tid_sid_invoker] [DockerForMacClient] running /usr/local/bin/docker ps --quiet --no-trunc --all --filter name=wsk0_ (timeout: 1 minute) [marker:invoker_docker.ps_start:4841]
[2019-06-24T17:22:31.265Z] [INFO] [#tid_sid_invoker] [DockerForMacClient]  [marker:invoker_docker.ps_finish:4927:86]
[2019-06-24T17:22:31.289Z] [INFO] [#tid_sid_invoker] [DockerContainerFactory] removing 0 action containers.
[2019-06-24T17:22:31.502Z] [INFO] [#tid_sid_unknown] [MemoryArtifactStore] Created MemoryStore for [activations]
[2019-06-24T17:22:31.650Z] [INFO] [#tid_sid_invoker] [MemoryArtifactStore] [QUERY] 'subjects' searching 'namespaceThrottlings/blockedNamespaces [marker:database_queryView_start:5312]
[2019-06-24T17:22:31.659Z] [INFO] [#tid_sid_invoker] [MemoryArtifactStore]  [marker:database_queryView_finish:5322:9]
[2019-06-24T17:22:31.664Z] [INFO] [#tid_sid_unknown] [InvokerReactive] updated blacklist to 0 entries
[2019-06-24T17:22:31.879Z] [INFO] [#tid_sid_dispatcher] [MessageFeed] handler capacity = 8, pipeline fill at = 8, pipeline depth = 16
[2019-06-24T17:22:31.900Z] [INFO] [#tid_sid_invokerWarmup] [ContainerPool] pre-warming 2 nodejs:10 256 MB
[2019-06-24T17:22:31.901Z] [INFO] [#tid_sid_controller] [Controller] loadbalancer initialized: LeanBalancer
[2019-06-24T17:22:31.997Z] [INFO] [#tid_sid_controller] [KindRestrictor] all kinds are allowed, the white-list is not specified
[2019-06-24T17:22:32.209Z] [INFO] [#tid_sid_invokerWarmup] [DockerForMacClient] running /usr/local/bin/docker run -d -p 0:8080 --cpu-shares 256 --memory 256m --memory-swap 256m --network bridge -e __OW_API_HOST=http://host.docker.internal:3233 --name wsk0_1_prewarm_nodejs10 --cap-drop NET_RAW --cap-drop NET_ADMIN --ulimit nofile=1024:1024 --pids-limit 1024 --log-driver json-file openwhisk/action-nodejs-v10:latest (timeout: 1 minute) [marker:invoker_docker.run_start:5871]
[2019-06-24T17:22:32.210Z] [INFO] [#tid_sid_invokerWarmup] [DockerForMacClient] running /usr/local/bin/docker run -d -p 0:8080 --cpu-shares 256 --memory 256m --memory-swap 256m --network bridge -e __OW_API_HOST=http://host.docker.internal:3233 --name wsk0_2_prewarm_nodejs10 --cap-drop NET_RAW --cap-drop NET_ADMIN --ulimit nofile=1024:1024 --pids-limit 1024 --log-driver json-file openwhisk/action-nodejs-v10:latest (timeout: 1 minute) [marker:invoker_docker.run_start:5871]
[2019-06-24T17:22:32.619Z] [INFO] [#tid_sid_unknown] [ExtendedDockerClient] Detected docker client version 18.05.0-ce
[2019-06-24T17:22:34.085Z] [INFO] [#tid_sid_controller] [ActionsApi] actionSequenceLimit '50'
[2019-06-24T17:23:32.301Z] [ERROR] [#tid_sid_invokerWarmup] [DockerForMacClient] info: command was terminated, took longer than 1 minute, code: 143 (terminated by signal SIGTERM), stdout: , stderr: Unable to find image 'openwhisk/action-nodejs-v10:latest' locally
latest: Pulling from openwhisk/action-nodejs-v10
c5e155d5a1d1: Pulling fs layer
221d80d00ae9: Pulling fs layer
4250b3117dca: Pulling fs layer
3b7ca19181b2: Pulling fs layer
425d7b2a5bcc: Pulling fs layer
69df12c70287: Pulling fs layer
ad53476a61f2: Pulling fs layer
204bb8bac4a1: Pulling fs layer
e15197924b53: Pulling fs layer
b095df4a85d6: Pulling fs layer
9038287d9c58: Pulling fs layer
3a303a4e5699: Pulling fs layer
c3b8ce658191: Pulling fs layer
204bb8bac4a1: Waiting
e15197924b53: Waiting
b095df4a85d6: Waiting
9038287d9c58: Waiting
3a303a4e5699: Waiting
c3b8ce658191: Waiting
ad53476a61f2: Waiting
3b7ca19181b2: Waiting
425d7b2a5bcc: Waiting
69df12c70287: Waiting
4250b3117dca: Verifying Checksum
4250b3117dca: Download complete
221d80d00ae9: Verifying Checksum
221d80d00ae9: Download complete
c5e155d5a1d1: Verifying Checksum
c5e155d5a1d1: Download complete
69df12c70287: Verifying Checksum
69df12c70287: Download complete
3b7ca19181b2: Verifying Checksum
3b7ca19181b2: Download complete
204bb8bac4a1: Download complete
c5e155d5a1d1: Pull complete
221d80d00ae9: Pull complete
ad53476a61f2: Verifying Checksum
ad53476a61f2: Download complete
4250b3117dca: Pull complete
b095df4a85d6: Verifying Checksum
b095df4a85d6: Download complete
e15197924b53: Verifying Checksum
e15197924b53: Download complete
9038287d9c58: Verifying Checksum
9038287d9c58: Download complete
3a303a4e5699: Verifying Checksum
3a303a4e5699: Download complete
c3b8ce658191: Download complete
3b7ca19181b2: Pull complete [marker:invoker_docker.run_error:65956:60072]
```
3. Pull nodejs runtime image on host
4. Create action hello.js

```bash
[2019-06-24T17:25:11.991Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] PUT /api/v1/namespaces/_/actions/hello overwrite=false
[2019-06-24T17:25:12.095Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [BasicAuthenticationDirective] authenticate: 23bc46b1-71f6-4ed5-8c54-816aa4f8c502
[2019-06-24T17:25:12.228Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [Identity] [GET] serving from datastore: CacheKey(23bc46b1-71f6-4ed5-8c54-816aa4f8c502) [marker:database_cacheMiss_counter:250]
[2019-06-24T17:25:12.228Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [MemoryArtifactStore] [QUERY] 'subjects' searching 'subjects/identities [marker:database_queryView_start:250]
[2019-06-24T17:25:12.236Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [MemoryArtifactStore] [GET] 'subjects' finding document: 'guest/limits' [marker:database_getDocument_start:258]
[2019-06-24T17:25:12.237Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [MemoryArtifactStore]  [marker:database_getDocument_finish:259:1]
[2019-06-24T17:25:12.245Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [MemoryArtifactStore]  [marker:database_queryView_finish:267:16]
[2019-06-24T17:25:12.455Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [WhiskAction] [GET] serving from datastore: CacheKey(guest/hello) [marker:database_cacheMiss_counter:475]
[2019-06-24T17:25:12.456Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [MemoryArtifactStore] [GET] 'whisks' finding document: 'id: guest/hello' [marker:database_getDocument_start:478]
[2019-06-24T17:25:12.457Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [MemoryArtifactStore]  [marker:database_getDocument_finish:478:0]
[2019-06-24T17:25:12.474Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [WhiskAction] invalidating CacheKey(guest/hello)
[2019-06-24T17:25:12.637Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [WhiskAction] write initiated on new cache entry
[2019-06-24T17:25:12.851Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [MemoryArtifactStore] [PUT] 'whisks' saving document: 'id: guest/hello, rev: null' [marker:database_saveDocument_start:869]
[2019-06-24T17:25:12.852Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [MemoryArtifactStore]  [marker:database_saveDocument_finish:870:1]
[2019-06-24T17:25:12.878Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [WhiskAction] write all done, caching CacheKey(guest/hello) Cached
[2019-06-24T17:25:12.990Z] [INFO] [#tid_Z3pKiYQ84Kz6QCzA81tcQVIXfBp7AwD4] [BasicHttpService] [marker:http_put.200_counter:1010:1010]```

5. Invoke function in blocking mode and you can see the warmup docker instance running. 

```bash
[2019-06-24T17:29:57.270Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] POST /api/v1/namespaces/_/actions/hello blocking=true&result=true
[2019-06-24T17:29:57.278Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [BasicAuthenticationDirective] authenticate: 23bc46b1-71f6-4ed5-8c54-816aa4f8c502
[2019-06-24T17:29:57.282Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [Identity] [GET] serving from cache: CacheKey(23bc46b1-71f6-4ed5-8c54-816aa4f8c502) [marker:database_cacheHit_counter:15]
[2019-06-24T17:29:57.326Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [WhiskActionMetaData] [GET] serving from cache: CacheKey(guest/hello) [marker:database_cacheHit_counter:58]
[2019-06-24T17:29:57.332Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [ActionsApi]  [marker:controller_blockingActivation_start:64]
[2019-06-24T17:29:57.333Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [ActionsApi] action activation id: c23b347a16034e9cbb347a16030e9cb0 [marker:controller_loadbalancer_start:65]
[2019-06-24T17:29:57.334Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [LeanBalancer] posting topic 'invoker0' with activation id 'c23b347a16034e9cbb347a16030e9cb0' [marker:controller_kafka_start:67]
[2019-06-24T17:29:57.341Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [LeanBalancer] posted to invoker0[0][-1] [marker:controller_kafka_finish:72:5]
[2019-06-24T17:29:57.341Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [ActionsApi]  [marker:controller_loadbalancer_finish:74:9]
[2019-06-24T17:29:57.360Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [InvokerReactive]  [marker:invoker_activation_start:82]
[2019-06-24T17:29:57.360Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [WhiskAction] [GET] serving from cache: CacheKey(guest/hello) [marker:database_cacheHit_counter:83]
[2019-06-24T17:29:57.360Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [ContainerPool] containerStart containerState: cold container: None activations: 1 of max 1 action: hello namespace: guest activationId: c23b347a16034e9cbb347a16030e9cb0 [marker:invoker_containerStart.cold_counter:90]
[2019-06-24T17:29:57.363Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [DockerForMacClient] running /usr/local/bin/docker run -d -p 0:8080 --cpu-shares 256 --memory 256m --memory-swap 256m --network bridge -e __OW_API_HOST=http://host.docker.internal:3233 --name wsk0_4_guest_hello --cap-drop NET_RAW --cap-drop NET_ADMIN --ulimit nofile=1024:1024 --pids-limit 1024 --log-driver json-file openwhisk/action-nodejs-v10:latest (timeout: 1 minute) [marker:invoker_docker.run_start:95]
[2019-06-24T17:29:58.358Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [DockerForMacClient]  [marker:invoker_docker.run_finish:1091:995]
[2019-06-24T17:29:58.376Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [DockerForMacClient] running /usr/local/bin/docker inspect --format {{(index (index .NetworkSettings.Ports "8080/tcp") 0).HostPort}} 1ae1f28d23b5ee334bb7916592da0e895bdfd2d1b309ae4429d1f70324e41976 (timeout: 10 seconds) [marker:invoker_docker.inspect_start:1108]
[2019-06-24T17:29:58.528Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [DockerForMacClient]  [marker:invoker_docker.inspect_finish:1260:152]
[2019-06-24T17:29:58.561Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [DockerContainer] sending initialization to ContainerId(1ae1f28d23b5ee334bb7916592da0e895bdfd2d1b309ae4429d1f70324e41976) ContainerAddress(localhost,32792) [marker:invoker_activationInit_start:1293]
[2019-06-24T17:30:00.705Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [DockerContainer] initialization result: ok [marker:invoker_activationInit_finish:3438:2137]
[2019-06-24T17:30:00.710Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [DockerContainer] sending arguments to /guest/hello at ContainerId(1ae1f28d23b5ee334bb7916592da0e895bdfd2d1b309ae4429d1f70324e41976) ContainerAddress(localhost,32792) [marker:invoker_activationRun_start:3442]
[2019-06-24T17:30:00.914Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [DockerContainer] running result: ok [marker:invoker_activationRun_finish:3646:202]
[2019-06-24T17:30:00.920Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [ContainerProxy]  [marker:invoker_collectLogs_start:3652]
[2019-06-24T17:30:00.922Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [ExtendedDockerClient] running /usr/local/bin/docker logs 1ae1f28d23b5ee334bb7916592da0e895bdfd2d1b309ae4429d1f70324e41976 --since 1561377598 --until 1561377601 --timestamps (timeout: 2 seconds) [marker:invoker_docker.logs_start:3655]
[2019-06-24T17:30:00.944Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [InvokerReactive] posted completion of activation c23b347a16034e9cbb347a16030e9cb0
[2019-06-24T17:30:00.952Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [LeanBalancer] received result ack for 'c23b347a16034e9cbb347a16030e9cb0'
[2019-06-24T17:30:00.952Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [ActionsApi]  [marker:controller_blockingActivation_finish:3685:3621]
[2019-06-24T17:30:00.957Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [BasicHttpService] [marker:http_post.200_counter:3689:3689]
[2019-06-24T17:30:01.726Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [ExtendedDockerClient]  [marker:invoker_docker.logs_finish:4459:804]
[2019-06-24T17:30:01.730Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [ContainerProxy]  [marker:invoker_collectLogs_finish:4462:809]
[2019-06-24T17:30:01.736Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [MemoryArtifactStore] [PUT] 'activations' saving document: 'id: guest/c23b347a16034e9cbb347a16030e9cb0, rev: null' [marker:database_saveDocument_start:4468]
[2019-06-24T17:30:01.739Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [MemoryArtifactStore]  [marker:database_saveDocument_finish:4472:3]
[2019-06-24T17:30:01.739Z] [INFO] [#tid_bRZrvyAOj8ps0iamWmSfvGQ80tbLbUFI] [InvokerReactive] posted completion of activation c23b347a16034e9cbb347a16030e9cb0
```

6. Wait for docker unpause to happen after 10 mins. 

```bash
[2019-06-24T17:40:23.476Z] [INFO] [#tid_sid_invokerNanny] [DockerForMacClient] running /usr/local/bin/docker unpause 1ae1f28d23b5ee334bb7916592da0e895bdfd2d1b309ae4429d1f70324e41976 (timeout: 10 seconds) [marker:invoker_docker.unpause_start:1077136]
[2019-06-24T17:40:23.733Z] [INFO] [#tid_sid_invokerNanny] [DockerForMacClient]  [marker:invoker_docker.unpause_finish:1077396:258]
[2019-06-24T17:40:23.797Z] [INFO] [#tid_sid_invokerNanny] [DockerForMacClient] running /usr/local/bin/docker rm -f 1ae1f28d23b5ee334bb7916592da0e895bdfd2d1b309ae4429d1f70324e41976 (timeout: 1 minute) [marker:invoker_docker.rm_start:1077459]
[2019-06-24T17:40:25.047Z] [INFO] [#tid_sid_invokerNanny] [DockerForMacClient]  [marker:invoker_docker.rm_finish:1078710:1251]
```

7. Again invoke action, it fails with below error.

```bash
[2019-06-24T17:40:50.818Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] POST /api/v1/namespaces/_/actions/hello blocking=true&result=true
[2019-06-24T17:40:50.835Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [BasicAuthenticationDirective] authenticate: 23bc46b1-71f6-4ed5-8c54-816aa4f8c502
[2019-06-24T17:40:50.841Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [Identity] [GET] serving from datastore: CacheKey(23bc46b1-71f6-4ed5-8c54-816aa4f8c502) [marker:database_cacheMiss_counter:29]
[2019-06-24T17:40:50.843Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [MemoryArtifactStore] [QUERY] 'subjects' searching 'subjects/identities [marker:database_queryView_start:29]
[2019-06-24T17:40:50.858Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [MemoryArtifactStore] [GET] 'subjects' finding document: 'guest/limits' [marker:database_getDocument_start:45]
[2019-06-24T17:40:50.858Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [MemoryArtifactStore]  [marker:database_getDocument_finish:45:0]
[2019-06-24T17:40:50.862Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [MemoryArtifactStore]  [marker:database_queryView_finish:48:18]
[2019-06-24T17:40:50.901Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [WhiskActionMetaData] [GET] serving from datastore: CacheKey(guest/hello) [marker:database_cacheMiss_counter:89]
[2019-06-24T17:40:50.901Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [MemoryArtifactStore] [GET] 'whisks' finding document: 'id: guest/hello' [marker:database_getDocument_start:90]
[2019-06-24T17:40:50.904Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [MemoryArtifactStore]  [marker:database_getDocument_finish:92:2]
[2019-06-24T17:40:50.930Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [ActionsApi]  [marker:controller_blockingActivation_start:119]
[2019-06-24T17:40:50.931Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [ActionsApi] action activation id: 3d5b132d251042e39b132d2510b2e311 [marker:controller_loadbalancer_start:119]
[2019-06-24T17:40:50.931Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [LeanBalancer] posting topic 'invoker0' with activation id '3d5b132d251042e39b132d2510b2e311' [marker:controller_kafka_start:120]
[2019-06-24T17:40:50.932Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [LeanBalancer] posted to invoker0[0][-1] [marker:controller_kafka_finish:121:1]
[2019-06-24T17:40:50.933Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [ActionsApi]  [marker:controller_loadbalancer_finish:122:3]
[2019-06-24T17:40:50.936Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [InvokerReactive]  [marker:invoker_activation_start:124]
[2019-06-24T17:40:50.963Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [WhiskAction] [GET] serving from datastore: CacheKey(guest/hello) [marker:database_cacheMiss_counter:126]
[2019-06-24T17:40:50.963Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [MemoryArtifactStore] [GET] 'whisks' finding document: 'id: guest/hello, rev: sha256-f9d277cdcb3b0291e51a5f5483c51fac211543345be2513e5235c7ed9993f02c' [marker:database_getDocument_start:126]
[2019-06-24T17:40:50.963Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [MemoryArtifactStore]  [marker:database_getDocument_finish:126:0]
[2019-06-24T17:40:50.963Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [MemoryArtifactStore] [ATT_GET] 'whisks' finding attachment 'mem:ZnVuY3Rpb24gbWFpbihhcmdzKSB7CiAgICAgICAgY29uc29sZS5sb2coYXJncykKCXJldHVybiB7cGF5bG9hZDogJ0hlbGxvICcgKyBhcmdzLm5hbWV9Owp9Cg==' of document 'id: guest/hello, rev: sha256-f9d277cdcb3b0291e51a5f5483c51fac211543345be2513e5235c7ed9993f02c' [marker:database_getDocumentAttachment_start:139]
[2019-06-24T17:40:51.013Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [WhiskAction] write initiated on existing cache entry, invalidating CacheKey(guest/hello), tid 4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY, state WriteInProgress
[2019-06-24T17:40:51.013Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [WhiskAction] write all done, caching CacheKey(guest/hello) Cached
[2019-06-24T17:40:51.018Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [ContainerPool] containerStart containerState: cold container: None activations: 1 of max 1 action: hello namespace: guest activationId: 3d5b132d251042e39b132d2510b2e311 [marker:invoker_containerStart.cold_counter:206]
[2019-06-24T17:40:51.020Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [DockerForMacClient] running /usr/local/bin/docker run -d -p 0:8080 --cpu-shares 256 --memory 256m --memory-swap 256m --network bridge -e __OW_API_HOST=http://host.docker.internal:3233 --name wsk0_5_guest_hello --cap-drop NET_RAW --cap-drop NET_ADMIN --ulimit nofile=1024:1024 --pids-limit 1024 --log-driver json-file openwhisk/action-nodejs-v10:latest (timeout: 1 minute) [marker:invoker_docker.run_start:209]
[2019-06-24T17:40:53.027Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [DockerForMacClient]  [marker:invoker_docker.run_finish:2216:2006]
[2019-06-24T17:40:53.027Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [DockerForMacClient] running /usr/local/bin/docker inspect --format {{(index (index .NetworkSettings.Ports "8080/tcp") 0).HostPort}} 6b1bfecca07499c7a3317d9240c69bfe3004e806579485e37cd9f90c9d5107aa (timeout: 10 seconds) [marker:invoker_docker.inspect_start:2216]
[2019-06-24T17:40:53.167Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [DockerForMacClient]  [marker:invoker_docker.inspect_finish:2356:140]
[2019-06-24T17:40:53.168Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [DockerContainer] sending initialization to ContainerId(6b1bfecca07499c7a3317d9240c69bfe3004e806579485e37cd9f90c9d5107aa) ContainerAddress(localhost,32793) [marker:invoker_activationInit_start:2356]
[2019-06-24T17:41:05.241Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] **[DockerContainer] initialization result: ConnectionError(org.apache.http.NoHttpResponseException: The target server failed to respond) [marker:invoker_activationInit_finish:14429:33]**
[2019-06-24T17:41:05.264Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [ContainerProxy]  [marker:invoker_collectLogs_start:14452]
[2019-06-24T17:41:05.264Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [ExtendedDockerClient] running /usr/local/bin/docker logs 6b1bfecca07499c7a3317d9240c69bfe3004e806579485e37cd9f90c9d5107aa --since 1561378253 --until 1561378254 --timestamps (timeout: 2 seconds) [marker:invoker_docker.logs_start:14452]
[2019-06-24T17:41:05.267Z] [INFO] [#tid_4aNUR6aIFQWK1jE0fbFZI6qt9k40DQsY] [InvokerReactive] posted completion of activation 3d5b132d251042e39b132d2510b2e311
```

chetanmeh commented:
@neerajmangal Thanks for reporting the details. I have now introduced a new `StandaloneDockerContainerFactory` which would now do couple of things

1. Instantiate an OS specific `DockerClient` implementation which takes care of any OS specific quirks related to launching Docker container
2. It would also do a pull for images which are considered to be provided by OpenWhisk. This info would be cached such that for subsequent calls no pull call is made

I tested it locally by first purging all nodejs runtimes

```bash
docker rmi $(docker images --filter=reference='openwhisk/*node*' -q)
```

Then post startup I can see in logs that pull is performed

```
[2019-06-25T15:45:12.404Z] [INFO] [#tid_sid_unknown] [ExtendedDockerClient] Detected docker client version 18.09.2
[2019-06-25T15:45:12.483Z] [INFO] [#tid_sid_invokerWarmup] [DockerForMacClient] running /usr/local/bin/docker pull openwhisk/action-nodejs-v10:latest (timeout: 10 minutes) [marker:invoker_docker.pull_start:2294]
[2019-06-25T15:45:12.483Z] [INFO] [#tid_sid_invokerWarmup] [DockerForMacClient] running /usr/local/bin/docker pull openwhisk/action-nodejs-v10:latest (timeout: 10 minutes) [marker:invoker_docker.pull_start:2294]
[2019-06-25T15:45:12.942Z] [INFO] [#tid_sid_controller] [ActionsApi] actionSequenceLimit '50'
[2019-06-25T15:45:52.378Z] [INFO] [#tid_sid_invokerWarmup] [DockerForMacClient]  [marker:invoker_docker.pull_finish:42189:39895]
[2019-06-25T15:45:52.379Z] [INFO] [#tid_sid_invokerWarmup] [DockerForMacClient]  [marker:invoker_docker.pull_finish:42189:39895]
[2019-06-25T15:45:52.380Z] [INFO] [#tid_sid_unknown] [StandaloneDockerContainerFactory] Pulled OpenWhisk provided image wsk0_2_prewarm_nodejs10
[2019-06-25T15:45:52.380Z] [INFO] [#tid_sid_unknown] [StandaloneDockerContainerFactory] Pulled OpenWhisk provided image wsk0_1_prewarm_nodejs10
[2019-06-25T15:45:52.450Z] [INFO] [#tid_sid_invokerWarmup] [DockerForMacClient] running /usr/local/bin/docker run -d -p 0:8080 --cpu-shares 256 --memory 256m --memory-swap 256m --network bridge -e __OW_API_HOST=http://host.docker.internal:3233 --name wsk0_2_prewarm_nodejs10 --cap-drop NET_RAW --cap-drop NET_ADMIN --ulimit nofile=1024:1024 --pids-limit 1024 --log-driver json-file openwhisk/action-nodejs-v10:latest (timeout: 1 minute) [marker:invoker_docker.run_start:42261]
[2019-06-25T15:45:52.450Z] [INFO] [#tid_sid_invokerWarmup] [DockerForMacClient] running /usr/local/bin/docker run -d -p 0:8080 --cpu-shares 256 --memory 256m --memory-swap 256m --network bridge -e __OW_API_HOST=http://host.docker.internal:3233 --name wsk0_1_prewarm_nodejs10 --cap-drop NET_RAW --cap-drop NET_ADMIN --ulimit nofile=1024:1024 --pids-limit 1024 --log-driver json-file openwhisk/action-nodejs-v10:latest (timeout: 1 minute) [marker:invoker_docker.run_start:42261]
[2019-06-25T15:45:53.389Z] [INFO] [#tid_sid_invokerWarmup] [DockerForMacClient]  [marker:invoker_docker.run_finish:43201:940]


```

rabbah commented:
Two suggestions: when startup detects that `wsk` is not configured properly, it will emit an `ERROR` which I think should be a `WARN` instead. I would also emit the `wsk property set ...` command as part of the same line if possible and color the whole text so it's easier to spot. It's easy to miss with text scrolling by.

rabbah commented:
if the runtimes file is missing, you get a stack trace instead of a nice error - i can open this is as a defect post merge i imagine there could be a number of things to improve as this is more heavily used.


```
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:47)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:86)
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:50)
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:51)
Caused by: java.io.FileNotFoundException: File 'runtimes.json' does not exist
	at org.apache.commons.io.FileUtils.openInputStream(FileUtils.java:297)
	at org.apache.commons.io.FileUtils.readFileToString(FileUtils.java:1805)
	at org.apache.openwhisk.standalone.StandaloneOpenWhisk$.configureRuntimeManifest(StandaloneOpenWhisk.scala:162)
	at org.apache.openwhisk.standalone.StandaloneOpenWhisk$.initialize(StandaloneOpenWhisk.scala:123)
	at org.apache.openwhisk.standalone.StandaloneOpenWhisk$.main(StandaloneOpenWhisk.scala:109)
	at org.apache.openwhisk.standalone.StandaloneOpenWhisk.main(StandaloneOpenWhisk.scala)
```

rabbah commented:
@chetanmeh i think the docker pull logic is too eager - it doesn't permit the use of local images that aren't on dockerhub (in a normal deployment we white list certain image prefixes to prevent the explicit docker pull). I suspect we can replicate this here, taking a look.

rabbah commented on core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/docker/StandaloneDockerContainerFactory.scala:
```diff
@@ -0,0 +1,100 @@
+    //For standard usage its expected that standard images have already been pulled in.
+    val imageName = actionImage.localImageName(runtimesRegistryConfig.url)
+    val pulled = if (!userProvidedImage && !pulledImages.contains(imageName)) {
+      docker.pull(imageName)(tid).map { _ =>
```
this will be problematic if one wants to use a build images that's not on dockerhub yet. 
a docker run will pull a missing image if necessary, maybe we can support the `bypass` options available on the invoker role today to skip pulling images. wdyt?

rabbah commented:
i would also like to add a `wskdev` target for standalone.

rabbah commented on core/standalone/README.md:
````diff
@@ -0,0 +1,153 @@
+$ ./gradlew :core:standalone:build
+```
+
+This would create the runnable jar in `core/standalone/build/libs/` directory
````
maybe `gradle ...:run` to then start the built jar?

chetanmeh replied:
Good point. Now you run the jar directly via `$ ./gradlew :core:standalone:run`

chetanmeh commented:
> if the runtimes file is missing, you get a stack trace instead of a nice error

Now it would show an error message like below

```
[openwhisk] Error: Validation failure for 'manifest' option parameters: /path/to/non-existing.json
```

chetanmeh commented:
> Two suggestions: when startup detects that `wsk` is not configured properly, it will emit an `ERROR` which I think should be a `WARN` instead. I would also emit the `wsk property set ...` command as part of the same line if possible and color the whole text so it's easier to spot. It's easy to miss with text scrolling by.

Now it marks it as warning

<img width="663" alt="image" src="https://user-images.githubusercontent.com/664531/60111466-9f472900-978b-11e9-95b8-f436fe8ed615.png">

chetanmeh commented:
> i think the docker pull logic is too eager - it doesn't permit the use of local images that aren't on dockerhub

Fixed that now by ensuring that eager pull is only done for `openwhisk/xxx` images

chetanmeh commented:
> i would also like to add a `wskdev` target for standalone.

Added `wskdev standalone` target now which directly launches the standalone jar via gradle

rabbah commented on core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/docker/DockerCliLogStore.scala:
```diff
@@ -0,0 +1,81 @@
+  private val logTimeSpanMargin = 1.second
+
+  def collectLogs(id: ContainerId, since: Instant, untill: Instant)(implicit transid: TransactionId): Future[String] = {
+    //Add a slight buffer to account for delay writes of logs
```
the logs are missing the stream information `stdout|stderr` -- discovered this by accident because `wsk activation logs --strip` won't strip the timestamp and stream correctly otherwise.

rabbah replied:
I adapted the cli https://github.com/apache/incubator-openwhisk-cli/pull/444.
